neqo-transport/src/connection.rs

```diff
@@ -2488,74 +2509,91 @@ impl SentPacket {
 
 #[derive(Debug, Default)]
 struct RttVals {
-    latest_rtt: u64,
-    smoothed_rtt: u64,
-    rttvar: u64,
-    min_rtt: u64,
-    max_ack_delay: u64,
+    latest_rtt: Duration,
+    smoothed_rtt: Option<Duration>,
+    rttvar: Duration,
+    min_rtt: Duration,
+    max_ack_delay: Duration,
 }
 
 impl RttVals {
-    fn update_rtt(&mut self, latest_rtt: u64, mut ack_delay: u64) {
+    fn update_rtt(&mut self, latest_rtt: Duration, ack_delay: Duration) {
         self.latest_rtt = latest_rtt;
         // min_rtt ignores ack delay.
         self.min_rtt = min(self.min_rtt, self.latest_rtt);
         // Limit ack_delay by max_ack_delay
-        ack_delay = min(ack_delay, self.max_ack_delay);
+        let ack_delay = min(ack_delay, self.max_ack_delay);
         // Adjust for ack delay if it's plausible.
         if self.latest_rtt - self.min_rtt > ack_delay {
             self.latest_rtt -= ack_delay;
         }
         // Based on {{?RFC6298}}.
-        if self.smoothed_rtt == 0 {
-            self.smoothed_rtt = self.latest_rtt;
-            self.rttvar = self.latest_rtt / 2;
-        } else {
-            let rttvar_sample = if self.smoothed_rtt > self.latest_rtt {
-                self.smoothed_rtt - self.latest_rtt
-            } else {
-                self.latest_rtt - self.smoothed_rtt
-            };
-            self.rttvar =
-                (3.0 / 4.0 * (self.rttvar as f64) + 1.0 / 4.0 * (rttvar_sample as f64)) as u64;
-            self.smoothed_rtt = (7.0 / 8.0 * (self.smoothed_rtt as f64)
-                + 1.0 / 8.0 * (self.latest_rtt as f64)) as u64;
+        match self.smoothed_rtt {
+            None => {
+                self.smoothed_rtt = Some(self.latest_rtt);
+                self.rttvar = self.latest_rtt / 2;
+            }
+            Some(smoothed_rtt) => {
+                let rttvar_sample = if smoothed_rtt > self.latest_rtt {
+                    smoothed_rtt - self.latest_rtt
+                } else {
+                    self.latest_rtt - smoothed_rtt
+                };
+
+                self.rttvar = Duration::from_micros(
+                    (3.0 / 4.0 * (self.rttvar.as_micros() as f64)
+                        + 1.0 / 4.0 * (rttvar_sample.as_micros() as f64))
```

Comment:
Ugh, we shouldn't need to use floating point here.  And it would be nice to have few uses of `as`.